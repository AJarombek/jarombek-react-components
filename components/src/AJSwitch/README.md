### Overview

Switch component that has two states - on and off.

**Use cases:**

```javascript
import { AJSwitch } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJSwitch
    onChange={state => console.info(`AJSwitch state: ${state}`)}
    disabled={false}
  />
```

### Files

| Filename                   | Description                                                                     |
|----------------------------|---------------------------------------------------------------------------------|
| `AJSwitch.js`              | The reusable `AJSwitch` component.                                              |
| `index.js`                 | Index file which exports the `AJSwitch` component.                              |
| `styles.js`                | JSS styles for the `AJSwitch` component.                                        |
