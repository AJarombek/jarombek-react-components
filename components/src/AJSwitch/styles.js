/**
 * Stylesheet for the AJSwitch component using JSS.
 * Author: Andrew Jarombek
 * Date: 12/31/2019
 */

export default {
  ajSwitch: {
    display: 'flex',
    flexDirection: 'row'
  },
  ajSwitchTail: {

  },
  ajSwitchHead: {

  }
};
