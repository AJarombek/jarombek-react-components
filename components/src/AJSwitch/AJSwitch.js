/**
 * Reusable switch that can be toggled between an 'on' and 'off' state.
 * @author Andrew Jarombek
 * @since 12/31/2019
 */

import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {createUseStyles} from 'react-jss';
import classnames from 'classnames';

import styles from './styles';

const useStyles = createUseStyles(styles);

/**
 * Component representing an off and on switch.  The user can plug in custom logic which will occur
 * when the off/on state changes.
 * @param onChange Function that is called when the switch is toggled.  The function takes a single
 * boolean argument, whose value is {@code true} if the switch is on and {@code false} if the
 * switch is off.
 * @param initialState The initial on/off state of the switch.
 * @param disabled Whether clicking on the switch changes its state.
 * @return {*} React elements representing a toggleable switch.
 */
const AJSwitch = ({onChange, initialState=false, disabled=false}) => {
  const classes = useStyles();
  const [state, setState] = useState(initialState);

  const onClick = () => {
    if (!disabled) {
      const newState = !state;
      setState(newState);
      onChange(newState);
    }
  };

  const mainClass = state ?
    disabled ?
      classnames(classes.ajSwitch, classes.ajSwitchActive, classes.ajSwitchDisabled, 'aj-switch'):
      classnames(classes.ajSwitch, classes.ajSwitchActive, 'aj-switch')
    :
    disabled ?
      classnames(classes.ajSwitch, classes.ajSwitchInactive, classes.ajSwitchDisabled, 'aj-switch'):
      classnames(classes.ajSwitch, classes.ajSwitchInactive, 'aj-switch');

  const headClass = state ?
    classnames(classes.ajSwitchHead, classes.ajSwitchHeadActive, 'aj-switch-head') :
    classnames(classes.ajSwitchHead, classes.ajSwitchHeadInactive, 'aj-switch-head');

  const tailClass = state ?
    classnames(classes.ajSwitchTail, classes.ajSwitchTailActive, 'aj-switch-tail') :
    classnames(classes.ajSwitchTail, classes.ajSwitchTailInactive, 'aj-switch-tail');

  return (
    <div className={mainClass} onClick={onClick}>
      <div className={headClass}> </div>
      <div className={tailClass}> </div>
    </div>
  );
};

AJSwitch.propTypes = {
  onChange: PropTypes.func.isRequired,
  initialState: PropTypes.bool,
  disabled: PropTypes.bool
};

export default AJSwitch;
