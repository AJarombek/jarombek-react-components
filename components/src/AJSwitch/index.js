import AJSwitch from './AJSwitch';

export default AJSwitch;
