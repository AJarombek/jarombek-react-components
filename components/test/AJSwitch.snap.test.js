import React from 'react';
import AJSwitch from '../src/AJSwitch/AJSwitch';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for the {@link ../../components/AJSwitch} component.
 * I hope your special day was wonderful.
 * @author Andrew Jarombek
 * @since 1/1/2020
 */

it('renders correctly', () => {
  const tree = renderer.create(<AJSwitch onChange={(f) => f} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with all props', () => {
  const tree = renderer.create(<AJSwitch onChange={(f) => f} initialState={true} disabled={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});
