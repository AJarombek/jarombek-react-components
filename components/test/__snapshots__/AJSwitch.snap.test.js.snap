// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ajSwitch-0-2-1 ajSwitchInactive-0-2-4 aj-switch"
  onClick={[Function]}
>
  <div
    className="ajSwitchHead-0-2-5 ajSwitchHeadInactive-0-2-7 aj-switch-head"
  >
     
  </div>
  <div
    className="ajSwitchTail-0-2-8 ajSwitchTailInactive-0-2-10 aj-switch-tail"
  >
     
  </div>
</div>
`;

exports[`renders correctly with all props 1`] = `
<div
  className="ajSwitch-0-2-1 ajSwitchActive-0-2-3 ajSwitchDisabled-0-2-2 aj-switch"
  onClick={[Function]}
>
  <div
    className="ajSwitchHead-0-2-5 ajSwitchHeadActive-0-2-6 aj-switch-head"
  >
     
  </div>
  <div
    className="ajSwitchTail-0-2-8 ajSwitchTailActive-0-2-9 aj-switch-tail"
  >
     
  </div>
</div>
`;
